src/data/ellpack_page.cu

```diff
@@ -12,14 +10,23 @@
 
 namespace xgboost {
 
-EllpackPage::EllpackPage(DMatrix* dmat) : impl_{new EllpackPageImpl(dmat)} {}
+EllpackPage::EllpackPage() : impl_{new EllpackPageImpl()} {}
+
+EllpackPage::EllpackPage(DMatrix* dmat, const BatchParam& param)
+    : impl_{new EllpackPageImpl(dmat, param)} {}
 
 EllpackPage::~EllpackPage() = default;
 
-EllpackPageImpl::EllpackPageImpl(DMatrix* dmat) : dmat_{dmat} {}
+size_t EllpackPage::Size() const {
+  return impl_->Size();
+}
+
+void EllpackPage::SetBaseRowId(size_t row_id) {
+  impl_->SetBaseRowId(row_id);
+}
 
 // Bin each input data entry, store the bin indices in compressed form.
-template<typename std::enable_if<true,  int>::type = 0>
+template<typename std::enable_if<true, int>::type = 0>
```

Comment:
Actually why is this line needed?

Reply:
Good question. :) I read this page: https://eli.thegreenplace.net/2014/sfinae-and-enable_if/, but still don't understand why it's here. Removed.

Reply:
i think that was added to prevent multiple symbol definition linker error. some of the tests includes the cuda file to test internal methods, and this kernel will be exposed as a global function with external linkage in multiple translation units. when they are put together to build a binary, linker will invariably complain.

making it `inline` or a function template will make the symbol visible only internally (internal linkage). it *could* have been made static as well to force an internal linkage; not sure why a function template was chosen.

this is moot presently, as with the ellpack refactorization, the internal kernel is moved to a separate file that isn't explicitly/implicitly included within the tests.